### ISSUE TITLE
JS exceptions thrown on filter update/compilation

### ISSUE THREAD

**Issue by crgmrgn** [AdguardTeam/AdGuardForSafari#525]
### Issue Details

* AdGuard for Safari version: 1.9.3beta (latest discreet download)
* Environment name and version: Safari 13.1.2 (13609.3.5.1.5)
* Operating System: macOS High Sierra 10.13.6
* Filters you use in AdGuard: Base, Tracking Protection
* Any other browser extensions you have installed: -

### Actual Behavior
Any update of the filters, whether thru timed periodic update or by toggling a filter on/off causes JS exceptions to be thrown and displayed as pop-up windows. Typically the execption is thrown 5x times, ie. 5x pop-ups

### Screenshots

<details><summary>Screenshot:</summary>

![Untitled](https://user-images.githubusercontent.com/48251/106807217-d1578700-6660-11eb-905e-7450a8056283.png)

</details>

**Comment by TheHasagi** [comment]
@crgmrgn 

Hello!

Does it happen every time you launch AdGuard for Safari? Could you please export logs and system info and send them to devteam@adguard.com with the subject "issue/525"?

**Comment by crgmrgn** [comment]
Yes, its consistent with every launch and can be triggered on demand by requesting any filter update, so seems to be related to each re-build of the rules db. The overall result is that nearly all blocking is ineffective as the ruleset is I presume never complete. 

I'll grab a logs zip later today when the system is to hand, as it was reverted to v1.8.10 to continue useful operation.

**Comment by crgmrgn** [comment]
We saw a very similar issue (#419) in an earlier beta, which was addressed by changes to SafariConverterLib, that fix seems to have maybe regressed in the v1.9.3beta again ...

**Comment by TheHasagi** [comment]
@crgmrgn 

That's something. Please, attach a crash report info to the logs.

**Comment by crgmrgn** [comment]
Couple of points:

1. Tried gathering logs (via menubar/GUI) on v1.9.3beta: whilst the log file zip is created (correct timestamp, etc) and populated with accurate info for the filters/json the internal log.log file is out of date and does not get regenerated (last internal timestamp is 2020) or captured to the zip.

Interestingly, the v1.8.10 MAS release on this same system captures a zip of the log files fine and its log.log reflects the correct runs of seconds before capture. 

Either the wrong file is being captured to the zip or the log is not getting written. What is the underlying location of the log.log file in the OS, maybe I can snag it manually and forward it with the rest of the log bundle?

2. Following the #419 thread, ConverterTool as supplied in v1.9.3beta does not successfully run on macOS 10.13.6 (attempting to execute it causes errors about missing Swift symbols/libs), hence I'd surmise that a failing run of ConverterTool is the failed pipe that is being reported in the dialogs ...

```
[~] > /Applications/Adguard\ for\ Safari.app/Contents/Resources/libs/ConverterTool
dyld: Symbol not found: _$s8RawValueSYTl
  Referenced from: /Applications/Adguard for Safari.app/Contents/Resources/libs/ConverterTool
  Expected in: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx/libswiftCore.dylib
 in /Applications/Adguard for Safari.app/Contents/Resources/libs/ConverterTool
Abort trap: 6
```

**Comment by Oleg-Chashko** [comment]
@crgmrgn 
Either the wrong file is being captured to the zip or the log is not getting written. What is the underlying location of the log.log file in the OS, maybe I can snag it manually and forward it with the rest of the log bundle?

/Users/User_Name/Library/Logs/AdGuardSafariApp

**Comment by crgmrgn** [comment]
Log capture ...
[adg_safari_logs_1612789021696.zip](https://github.com/AdguardTeam/AdGuardForSafari/files/5943825/adg_safari_logs_1612789021696.zip)

**Comment by crgmrgn** [comment]
If I substitute the `ConverterTool` binary from the v1.8.10 release for the binary that is distributed with v1.9.3beta, the JS exceptions go away and everything seems to be working correctly (from a users perspective), I don't know what other functionality might have been lost (or regressed), but its surely the root of the issue here …

The working binary is

```
[~] > md5 /Users/craig/Downloads/Adguard\ for\ Safari.app/Contents/Resources/libs/ConverterTool
MD5 (/Users/craig/Downloads/Adguard for Safari.app/Contents/Resources/libs/ConverterTool) = df81ef796618de64dfb6026d4f73d5b5
```

**Comment by crgmrgn** [comment]
Just installed the latest v1.9.5beta standalone release and attempted to verify operation, it shows the same failure modes and error messages as the earlier reported issues with v1.9.3beta … namely, each time the ruleset is converted, the system reports failed pipe/JS exception errors, as per the original screenshots.

This behaviour leaves the ruleset effectively neutered and hence the Safari session open to ads. 

I can attempt the work-around of reverting the `ConverterTool` binary, but I'm loath too as it will negate any enhancements/improvements made to its code in the last few newer releases.

**Comment by crgmrgn** [comment]
OK, I had some time to experiment today (v1.8.5beta on macOS 10.13.6) …

1. I substituted the v1.8.10 binary of `ConverterTool` (md5 above) with that supplied in the v1.8.5beta and can confirm that the JS exceptions and failing entries in the logs disappeared. The tool also reports v1.1.11, so is the code unchanged but its just build issues in the beta releases that inhibit operation?
2. v1.8.5beta "patched" with the above now functions *but* exhibits an old trait … the Safari AfS titlebar icon will disappear after a few seconds, then return, and repeat, until eventually it disappears for good. Checking "Preferences" for the extensions reveals only two extensions present "Adguard Safari Icon" and "Adguard Advanced Blocking" (both enabled). 
3. Interestingly, I note that if you clean start Safari and immediately open up the Prefs pane to extensions, you can observe 7x Adguard entries present and enabled, then watch each one disappear (over about ~10secs duration), until only the two remain, each disappearance co-inciding with the Safari Icon blinking out of view on the titlebar … I assume some background test/build/check is actually failing and Safari is retiring the extension … I'll take a look thru the system logs.

**Comment by Mizzick** [comment]
@crgmrgn 
The disappearance of extensions is a reliable sign of a broken application, I guess in fact it doesn't work at all. Have you noticed if this "patched" build still blocks some ads? Please have a log at system logs, there should be many errors listed.

Nonetheless, I want to say that we work on this issue right now and we will try to fix it in the next beta.

**Comment by crgmrgn** [comment]
Core blocking still seems to be functioning, maybe some of the more subtle rulesets aren't. With regard to errors in the system logs, there don't seem to be any significantly greater number than I recollect from previous released runs, but the more notable seem to be associated with the below:

```
error	11:42:27.675176 +0000	AdGuardForSafariExtension	Couldn't read values in CFPrefsPlistSource<0x6040002e3580> (Domain: TC3Q7MAJXF.com.adguard.safari.AdGuard, User: kCFPreferencesAnyUser, ByHost: Yes, Container: (null), Contents Need Refresh: Yes): Using kCFPreferencesAnyUser with a container is only allowed for System Containers, detaching from cfprefsd
```

This one is reported for `AdGuardForSafariExtension`, but I see similar constantly for all of the AdGuard extensions, I guess as they are all instances of one codebase.

**Comment by TheHasagi** [comment]
@crgmrgn 

Hello, can you please check once again with this special [build](https://uploads.adguard.com/myMoonk0n4e.zip)?

**Comment by crgmrgn** [comment]
That new special build is looking much better … 
- `ConverterTool` runs fine
- no JS exception pop-ups in AfS on ruleset updates
- Safari prefs shows 8x AdGuard extensions installed/enabled and they are stable
- all the relevant rules seem to be being applied as best I can ascertain
- titlebar icon is now stable (no flicker/disappearance)

Console messages show some more debug messages, so they've increased a touch in volume, but I assume thats during your test cycles only …

In short, on a quick acquaintance I'd personally be pretty happy running this going forward as a candidate for v1.9.x release ;)

**Comment by mrlife** [comment]
```
> error	11:42:27.675176 +0000	AdGuardForSafariExtension	Couldn't read values in CFPrefsPlistSource<0x6040002e3580> (Domain: TC3Q7MAJXF.com.adguard.safari.AdGuard, User: kCFPreferencesAnyUser, ByHost: Yes, Container: (null), Contents Need Refresh: Yes): Using kCFPreferencesAnyUser with a container is only allowed for System Containers, detaching from cfprefsd
```

Hi @crgmrgn, do you recall where in macOS you saw this error? I found it once when researching an issue but don't remember where. Thanks for any help!
